### DISCUSSION TITLE
UDP listener

### DISCUSSION THREAD

**Discussion by jjmSE** [akkadotnet/akka.net#5347] [Q&A]
Hello everyone!

We have a system to collect information from lots of devices (millions) and we use UDPConnected, so a connection is only able to send to the remoteAddress it was connected to, and will receive datagrams only from that address.

We also receive events from these devices so we have a listener, but can't use UDPConnected for that (Bind function is only available for UDP actor). We need to receive all the datagrams in different sockets and received a message when it is closed by the client (similar to TCP Bind). What is the best way to do that?

Thanks!!

**Comment by Aaronontheweb** [comment]
Hi @jjmSE - let me get back to you on this. Tied up with a long engagement this week but I will reply with some ideas.

**Reply by jjmSE** [reply]
Thanks, your help would be greatly appreciated.

**Reply by jjmSE** [reply]
Hi @Aaronontheweb, do you have any idea about this topic? Thanks.

**Reply by Aaronontheweb** [reply]
> We also receive events from these devices so we have a listener, but can't use UDPConnected for that (Bind function is only available for UDP actor). We need to receive all the datagrams in different sockets and received a message when it is closed by the client (similar to TCP Bind). What is the best way to do that?

I'm wondering if Akka.Streams.IO.Udp can help with this. I have an example of how to build a duplex TCP client <--> server that I updated for my dotnetconf talk last week: https://github.com/Aaronontheweb/Akka.Streams.Benchmark

Let me check and see how that could be done via UDP in Akka.Streams. This is all built on top of Akka.IO's Udp functionality.

**Reply by Aaronontheweb** [reply]
Well damn, we do not have a UDP version of this class available: https://github.com/akkadotnet/akka.net/blob/e27220a09ed10d71bc32cb73f4002641b0faaf94/src/core/Akka.Streams/Dsl/Tcp.cs#L24-L61 - so it looks like a built-in Akka.Streams solution may be off the table. Let me try something else....

**Reply by Aaronontheweb** [reply]
https://github.com/akkadotnet/akka.net/blob/e27220a09ed10d71bc32cb73f4002641b0faaf94/src/core/Akka.Tests/IO/UdpConnectedIntegrationSpec.cs#L41-L75

So just to make sure I understand why the `Udp.Bind` command with `UdpConnected` won't work for you, is it because you don't get notified if the UDP device goes offline? As in, there's no heartbeat detection? It seems like you just use the UDP actor with `UdpConnected` per the integration tests I linked - can you help me better understand why that won't work?

**Reply by ismaelhamed** [reply]
I was just looking into this yesterday. The problem is that being Udp a connectionless protocol, there's no way to reliably implement a "disconnected" mechanism. Even the `Connect()` at the Udp socket level merely sets the default destination address for send, and then tells the OS for which incoming address to actually accept packets. But there's no concept of connection, and obviously no handshake, confirmation, or any way to assess the state of the destination address.

Looks like you typically must build this on top of Udp, like Google's SIP protocol.

**Reply by Aaronontheweb** [reply]
You could use an Akka.Streams `KeepAlive` stage to maybe add a heartbeat mechanism over UDP on the server-side. That would work. Otherwise, you'd need a dead man's switch to do it while listening for incoming events from the downstream devices.

**Comment by Kiddinglife** [comment]
hey @jjmSE looks like an ingress issue. so basically you need write a udp ingress gateway for your devices traffics. I guess even akka.java does not have such built-in gateway. have you tried write a simple aka system that has two interfaces: udp interface for the incoming device data and the other kka.net actor for the outgoing data, with a simple protocol transformation from your custom protocol to akka-native protocol. if you are using k8s, you can use hostport for your ingress docker instance. and then syn the ingress instances to some backend so that your device connects and get the list of healthy ingress instances to do client-side load balance.
Aother similar solution is to directly connec to your actor system that contains your device actor if your actor system can be public-faced to avoid the extra costs to rum those ingress ctor systems.
